lib/backends/webgl/ops/conv-pack.ts

```diff
@@ -0,0 +1,117 @@
+      }
+    }
+    PoolConvUtil.adjustPadsBasedOnAutoPad(
+        inputs[0].dims, this.strides, this.dilations, this.kernelShape, this.pads, this.autoPad);
+    Logger.verbose(
+        'Conv',
+        `autpPad:${this.autoPad}, dilations:${this.dilations}, group:${this.group}, kernelShape:${
+            this.kernelShape}, pads:${this.pads}, strides:${this.strides}`);
+
+    const outputShape = WebGLConv.calcOutputShape(xshape, kshape, this.dilations, this.pads, this.strides);
+    const im2col = new WebGLIm2ColPacked(outputShape, kshape, this.dilations, this.pads, this.strides);
+    const matmul = new WebGLMatMulPacked();
+    const reshape = new WebGLReshapePacked();
+    // shape for kernel reshape
+    const shape = new Tensor([2], 'int32');
+    shape.data[0] = kshape[0];
+    shape.data[1] = kshape[1] * kshape[2] * kshape[3];
+
+    if (!this.artifacts) {
+      this.artifacts = [];
+      this.programInfo = [];
+      this.programInfo[0] = im2col.createProgramInfo(inferenceHandler, [inputs[0], inputs[1]]);
+      this.artifacts[0] = programManager.build(this.programInfo[0]);
+
+      this.programInfo[1] = reshape.createProgramInfo(inferenceHandler, [inputs[1], shape]);
+      this.artifacts[1] = programManager.build(this.programInfo[1]);
+    }
+
+    // run im2col
+    const runDataIm2col = im2col.createRunData(inferenceHandler, this.programInfo[0], [inputs[0], inputs[1]]);
+    inferenceHandler.checkAndUpdateTextureForm(this.artifacts[0], runDataIm2col);
+    programManager.run(this.artifacts[0], runDataIm2col);
+    const im2colOutput = runDataIm2col.outputTextureData.tensor;
+
+    // reshape kernel
+    const runDataKernelReshape = reshape.createRunData(inferenceHandler, this.programInfo[1], [inputs[1], shape]);
+    inferenceHandler.checkAndUpdateTextureForm(this.artifacts[1], runDataKernelReshape);
+    programManager.run(this.artifacts[1], runDataKernelReshape);
+    const kernelReshaped = runDataKernelReshape.outputTextureData.tensor;
+
+    // run matmul
+    const hasBias = (inputs.length === 3);
+    if (this.artifacts.length === 2) {
+      this.programInfo[2] = matmul.createProgramInfo(
+          inferenceHandler, hasBias ? [kernelReshaped, im2colOutput, inputs[2]] : [kernelReshaped, im2colOutput]);
+      this.artifacts[2] = programManager.build(this.programInfo[2]);
+    }
+    const runDataMatmul = matmul.createRunData(
+        inferenceHandler, this.programInfo[2],
+        hasBias ? [kernelReshaped, im2colOutput, inputs[2]] : [kernelReshaped, im2colOutput]);
+    inferenceHandler.checkAndUpdateTextureForm(this.artifacts[2], runDataMatmul);
+    programManager.run(this.artifacts[2], runDataMatmul);
+    const matmulOutput = runDataMatmul.outputTextureData.tensor;
+
+    // reshape output
+    const outputShapeTensor = new Tensor([outputShape.length], 'int32');
+
+    for (let i = 0; i < outputShape.length; i++) {
+      outputShapeTensor.data[i] = outputShape[i];
+    }
+
+    if (this.artifacts.length === 3) {
```

Comment:
same as above, at this point, this.artifacts.length needs to be 3. We should add an assert here.

Reply:
done